File: src/main/java/com/robohorse/robopojogenerator/generator/common/ClassField.java
package com.robohorse.robopojogenerator.generator.common;

import com.robohorse.robopojogenerator.generator.consts.templates.ArrayItemsTemplate;
import com.robohorse.robopojogenerator.generator.consts.ClassEnum;

/**
 * Created by vadim on 29.10.16.
 */
public class ClassField {
    private ClassEnum classEnum;
    private String className;
    private ClassField classField;

    public ClassField() {
    }

    public ClassField(ClassEnum classEnum) {
        this.classEnum = classEnum;
    }

    public ClassField(String className) {
        this.className = className;
    }

    public void setClassField(ClassField decorator) {
        if (classField == null) {
            classField = decorator;

        } else {
            classField.setClassField(decorator);
        }
    }

    public String getJavaItem() {
        return null != classField ? wrapListJava() :
                (null != className ? className : classEnum.getPrimitive());
    }

    public String getKotlinItem() {
        return null != classField ? wrapListKotlin() :
                (null != className ? className : classEnum.getKotlin());
    }

    private String wrapListJava() {
        return String.format(ArrayItemsTemplate.LIST_OF_ITEM, classField.getJavaBoxed());
    }

    private String wrapListKotlin() {
        return String.format(ArrayItemsTemplate.LIST_OF_ITEM, classField.getKotlinItem());
    }

    private String getJavaBoxed() {
        return null != classField ? wrapListJava() :
                (null != className ? className : classEnum.getBoxed());
    }
}

File: src/main/kotlin/com/robohorse/robopojogenerator/generator/consts/common/ClassField.kt
package com.robohorse.robopojogenerator.generator.consts.common

import com.robohorse.robopojogenerator.generator.consts.ClassEnum
import com.robohorse.robopojogenerator.generator.consts.templates.ArrayItemsTemplate

data class ClassField @JvmOverloads constructor(
        var classEnum: ClassEnum? = null,
        var className: String? = null,
        var classField: ClassField? = null
) {
    fun getJavaItem(primitive: Boolean = true) =
            if (primitive) {
                getJavaItemPrimitive()
            } else {
                getJavaBoxed()
            }

    private fun getJavaItemPrimitive(): String? {
        return if (null != classField) wrapListJava() else if (null != className) className else classEnum?.primitive
    }

    fun getKotlinItem(): String? {
        return if (null != classField) wrapListKotlin() else if (null != className) className else classEnum?.kotlin
    }

    private fun wrapListJava(): String? {
        return java.lang.String.format(ArrayItemsTemplate.LIST_OF_ITEM, classField?.getJavaBoxed())
    }

    private fun wrapListKotlin(): String? {
        return java.lang.String.format(ArrayItemsTemplate.LIST_OF_ITEM, classField?.getKotlinItem())
    }

    private fun getJavaBoxed(): String? {
        return if (null != classField) wrapListJava() else if (null != className) className else classEnum?.boxed
    }
}
